File: .py/114.FlattenBinaryTreeToLinkedList.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def flatten(self, root: Optional[TreeNode]) -> None:
        def flat(node: Optional[TreeNode]):
            if not node:
                return None

            left = node.left
            right = node.right

            if left:
                node.right = left
                node.left = None  
                current = node.right
                while current.right:
                    current = current.right
                current.right = right
            flat(node.right)

        flat(root)

File: .cpp/114.FlattenBinaryTreeToLinkedList.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    void flat(TreeNode *node)
    {
        if (!node)
            return;
        TreeNode *left = node->left;
        TreeNode *right = node->right;

        if (left)
        {
            node->right = left;
            node->left = NULL;
            TreeNode *current = node->right;
            while (current->right)
            {
                current = current->right;
            }
            current->right = right;
        }
        flat(node->right);
    }
    void flatten(TreeNode *root)
    {
        flat(root);
    }
};